File: tools/typescript/src/openai/toolkit.ts
import {ToolkitCore, ToolkitConfig} from '../shared/toolkit-core';
import type {
  ChatCompletionTool,
  ChatCompletionMessageToolCall,
  ChatCompletionToolMessageParam,
} from 'openai/resources';

class StripeAgentToolkit {
  private _core: ToolkitCore;
  private _tools: ChatCompletionTool[] = [];

  /**
   * The tools available in the toolkit.
   * @deprecated Access tools via getTools() after calling initialize().
   * Direct property access will return empty array if not initialized.
   */
  get tools(): ChatCompletionTool[] {
    this._core.warnIfNotInitialized();
    return this._tools;
  }

  constructor(config: ToolkitConfig) {
    this._core = new ToolkitCore(config);
  }

  /**
   * Initialize the toolkit by connecting to the MCP server.
   * Must be called before using tools.
   */
  initialize(): Promise<void> {
    return this._core.initialize((filteredTools) => {
      // Convert MCP tools to OpenAI ChatCompletionTool format
      // MCP already provides JSON Schema, which OpenAI expects
      this._tools = filteredTools.map((tool) => ({
        type: 'function' as const,
        function: {
          name: tool.name,
          description: tool.description || tool.name,
          parameters: tool.inputSchema || {type: 'object', properties: {}},
        },
      }));
    });
  }

  /**
   * Check if the toolkit has been initialized.
   */
  isInitialized(): boolean {
    return this._core.isInitialized();
  }

  /**
   * Get the tools in OpenAI ChatCompletionTool format.
   * Throws if not initialized.
   */
  getTools(): ChatCompletionTool[] {
    this._core.ensureInitialized();
    return this._tools;
  }

  /**
   * Processes a single OpenAI tool call by executing the requested function.
   *
   * @param {ChatCompletionMessageToolCall} toolCall - The tool call object from OpenAI containing
   *   function name, arguments, and ID.
   * @returns {Promise<ChatCompletionToolMessageParam>} A promise that resolves to a tool message
   *   object containing the result of the tool execution with the proper format for the OpenAI API.
   */
  async handleToolCall(
    toolCall: ChatCompletionMessageToolCall
  ): Promise<ChatCompletionToolMessageParam> {
    this._core.ensureInitialized();

    const args = JSON.parse(toolCall.function.arguments);
    const response = await this._core.stripe.run(toolCall.function.name, args);
    return {
      role: 'tool',
      tool_call_id: toolCall.id,
      content: response,
    } as ChatCompletionToolMessageParam;
  }

  /**
   * Close the toolkit connection and clean up resources.
   * Safe to call multiple times (idempotent).
   */
  close(): Promise<void> {
    return this._core.close(() => {
      this._tools = [];
    });
  }
}

/**
 * Factory function to create and initialize a StripeAgentToolkit.
 * Provides a simpler async initialization pattern.
 *
 * @example
 * const toolkit = await createStripeAgentToolkit({
 *   secretKey: 'rk_test_...',
 *   configuration: { actions: { customers: { create: true } } }
 * });
 * const tools = toolkit.getTools();
 */
export async function createStripeAgentToolkit(
  config: ToolkitConfig
): Promise<StripeAgentToolkit> {
  const toolkit = new StripeAgentToolkit(config);
  await toolkit.initialize();
  return toolkit;
}

export default StripeAgentToolkit;

File: tools/python/stripe_agent_toolkit/ag2/toolkit.py
"""Stripe Agent Toolkit for AG2 (formerly AutoGen)."""

import asyncio
import concurrent.futures
from typing import Any

from autogen.tools import Tool

from ..shared.toolkit_core import ToolkitCore
from ..shared.mcp_client import McpTool
from ..configuration import Configuration


class StripeAgentToolkit(ToolkitCore[list[Tool]]):
    """Stripe Agent Toolkit for AG2 (formerly AutoGen).

    Example:
        toolkit = await create_stripe_agent_toolkit(
            secret_key='rk_test_...',
        )
        tools = toolkit.get_tools()

        from autogen import ConversableAgent
        agent = ConversableAgent(name="billing", llm_config=llm_config)
        for tool in tools:
            tool.register_tool(agent)

        await toolkit.close()
    """

    def __init__(
        self,
        secret_key: str,
        configuration: Configuration | None = None,
    ):
        super().__init__(secret_key, configuration)
        self._executor = concurrent.futures.ThreadPoolExecutor(max_workers=4)

    def _empty_tools(self) -> list[Tool]:
        return []

    def _convert_tools(self, mcp_tools: list[McpTool]) -> list[Tool]:
        return [self._create_tool(t) for t in mcp_tools]

    def _create_tool(self, mcp_tool: McpTool) -> Tool:
        tool_name = mcp_tool["name"]
        description = mcp_tool.get("description", tool_name)
        run = self.run_tool
        executor = self._executor

        def call_stripe(**kwargs: Any) -> str:
            """Execute a Stripe tool via MCP."""
            try:
                asyncio.get_running_loop()
            except RuntimeError:
                return asyncio.run(run(tool_name, kwargs))
            # Already in async context — run in thread pool
            future = executor.submit(asyncio.run, run(tool_name, kwargs))
            return future.result()

        call_stripe.__name__ = tool_name
        call_stripe.__doc__ = description

        return Tool(
            name=tool_name,
            description=description,
            func_or_tool=call_stripe,
            parameters_json_schema=mcp_tool.get("inputSchema"),
        )

    async def close(self) -> None:
        """Close MCP connection and thread pool."""
        self._executor.shutdown(wait=False)
        await super().close()

    @property
    def tools(self) -> list[Tool]:
        """
        The tools available in the toolkit.

        .. deprecated::
            Access tools via get_tools() after calling initialize().
        """
        return self._get_tools_with_warning()


async def create_stripe_agent_toolkit(
    secret_key: str,
    configuration: Configuration | None = None,
) -> StripeAgentToolkit:
    """Create and initialize a StripeAgentToolkit for AG2.

    Args:
        secret_key: Stripe API key (rk_* recommended over sk_*)
        configuration: Optional Stripe configuration

    Returns:
        Initialized StripeAgentToolkit
    """
    toolkit = StripeAgentToolkit(secret_key, configuration)
    await toolkit.initialize()
    return toolkit
